src/darwin/Framework/CHIP/ServerEndpoint/MTRServerCluster.mm

```diff
@@ -0,0 +1,228 @@
+ *    limitations under the License.
+ */
+
+#import "MTRDefines_Internal.h"
+#import "MTRDeviceController_Internal.h"
+#import "MTRLogging_Internal.h"
+#import "MTRServerAttribute_Internal.h"
+#import "MTRServerCluster_Internal.h"
+#import "MTRServerEndpoint_Internal.h"
+#import <Matter/MTRClusterConstants.h>
+#import <Matter/MTRServerCluster.h>
+
+#include <app/clusters/descriptor/descriptor.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/support/SafeInt.h>
+
+using namespace chip;
+
+MTR_DIRECT_MEMBERS
+@implementation MTRServerCluster {
+    /**
+     * The access grants our API consumer can manipulate directly.  Never
+     * touched on the Matter queue.
+     */
+    NSMutableSet<MTRAccessGrant *> * _accessGrants;
+    NSMutableArray<MTRServerAttribute *> * _attributes;
+    MTRDeviceController * __weak _deviceController;
+}
+
+- (nullable instancetype)initWithClusterID:(NSNumber *)clusterID clusterRevision:(NSNumber *)clusterRevision
+{
+    auto clusterIDValue = clusterID.unsignedLongLongValue;
+    if (!CanCastTo<ClusterId>(clusterIDValue)) {
+        MTR_LOG_ERROR("MTRServerCluster provided too-large cluster ID: 0x%llx", clusterIDValue);
+        return nil;
+    }
+
+    auto id = static_cast<ClusterId>(clusterIDValue);
+    if (!IsValidClusterId(id)) {
+        MTR_LOG_ERROR("MTRServerCluster provided invalid cluster ID: 0x%" PRIx32, id);
+        return nil;
+    }
+
+    if (id == MTRClusterIDTypeDescriptorID) {
+        MTR_LOG_ERROR("Should be using initDescriptorCluster to initialize an MTRServerCluster for Descriptor");
+        return nil;
+    }
+
+    auto revisionValue = clusterRevision.unsignedLongLongValue;
+    if (revisionValue < 1 || revisionValue > 0xFFFF) {
+        MTR_LOG_ERROR("MTRServerCluster provided invalid cluster revision: 0x%llx", revisionValue);
+        return nil;
+    }
+
+    return [self initInternalWithClusterID:clusterID clusterRevision:clusterRevision accessGrants:[NSSet set] attributes:@[]];
+}
+
+- (instancetype)initDescriptorCluster
+{
+    return [self initInternalWithClusterID:@(MTRClusterIDTypeDescriptorID) clusterRevision:@(app::Clusters::Descriptor::kClusterRevision) accessGrants:[NSSet set] attributes:@[]];
+}
+
+- (instancetype)initInternalWithClusterID:(NSNumber *)clusterID clusterRevision:(NSNumber *)clusterRevision accessGrants:(NSSet *)accessGrants attributes:(NSArray *)attributes
+{
+    if (!(self = [super init])) {
+        return nil;
+    }
+
+    _clusterID = [clusterID copy];
+    _clusterRevision = [clusterRevision copy];
+    _accessGrants = [[NSMutableSet alloc] init];
+    _attributes = [[NSMutableArray alloc] init];
+    _matterAccessGrants = [NSSet set];
+    _parentEndpoint = kInvalidEndpointId;
+
+    for (MTRAccessGrant * grant in accessGrants) {
+        // Since our state is MTRServerEndpointStateInitializing, we know this
+        // will succeed.
+        [self addAccessGrant:[grant copy]];
```

Comment:
Probably no need to copy `grant` and `attr` here?

Reply:
Actually, `[attr copy]` is nonsensical now; I should take that out.  I guess we can just use the same grant without copying ok.

Reply:
Done.